File: learning-core/executable/retry/cpp/retry.cpp
#include <functional>
#include <stdexcept>
#include <string>
#include <unordered_map>

class transient_error : public std::runtime_error {
public:
    using std::runtime_error::runtime_error;
};

class permanent_error : public std::runtime_error {
public:
    using std::runtime_error::runtime_error;
};

std::string call_with_retry(const std::function<std::string()>& operation, int max_attempts) {
    if (max_attempts < 1) {
        throw std::invalid_argument("max_attempts must be >= 1");
    }

    for (int attempt = 1; attempt <= max_attempts; ++attempt) {
        try {
            return operation();
        } catch (const permanent_error&) {
            throw;
        } catch (const transient_error&) {
            if (attempt == max_attempts) {
                throw;
            }
        }
    }

    throw std::logic_error("unreachable");
}

class IdempotentStore {
public:
    std::string create_once(const std::string& key, const std::string& value) {
        if (const auto it = results_.find(key); it != results_.end()) {
            return it->second;
        }
        ++side_effects;
        results_[key] = value;
        return value;
    }

    int side_effects{0};

private:
    std::unordered_map<std::string, std::string> results_;
};

File: learning-core/executable/retry/python/retry.py
from __future__ import annotations

from collections.abc import Callable


class TransientError(RuntimeError):
    pass


class PermanentError(RuntimeError):
    pass


def call_with_retry(operation: Callable[[], str], *, max_attempts: int = 3) -> str:
    if max_attempts < 1:
        raise ValueError("max_attempts must be >= 1")

    last_error: TransientError | None = None
    for _ in range(max_attempts):
        try:
            return operation()
        except PermanentError:
            raise
        except TransientError as exc:
            last_error = exc

    assert last_error is not None
    raise last_error


class IdempotentStore:
    def __init__(self) -> None:
        self._results: dict[str, str] = {}
        self.side_effects = 0

    def create_once(self, key: str, value: str) -> str:
        if key in self._results:
            return self._results[key]
        self.side_effects += 1
        self._results[key] = value
        return value
